httpx/content.py

```diff
@@ -0,0 +1,162 @@
+    Request content encoded as plain bytes, using an async byte iterator.
+    """
+
+    def __init__(self, aiterator: typing.AsyncIterator[bytes]) -> None:
+        self.aiterator = aiterator
+
+    def can_rewind(self) -> bool:
+        return False
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        return {"Transfer-Encoding": "chunked"}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        async for part in self.aiterator:
+            yield part
+
+
+class JSONRequestContent(RequestContent):
+    """
+    Request content encoded as JSON.
+    """
+
+    def __init__(self, json: typing.Any) -> None:
+        self.body = json_dumps(json).encode("utf-8")
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        content_length = str(len(self.body))
+        content_type = "application/json"
+        return {"Content-Length": content_length, "Content-Type": content_type}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield self.body
+
+
+class URLEncodedRequestContent(RequestContent):
+    """
+    Request content as URL encoded form data.
+    """
+
+    def __init__(self, data: dict) -> None:
+        self.body = urlencode(data, doseq=True).encode("utf-8")
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        content_length = str(len(self.body))
+        content_type = "application/x-www-form-urlencoded"
+        return {"Content-Length": content_length, "Content-Type": content_type}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield self.body
+
+
+class MultipartRequestContent(RequestContent):
+    """
+    Request content as multipart encoded form data.
+    """
+
+    def __init__(self, data: dict, files: dict, boundary: bytes = None) -> None:
+        self.body, self.content_type = multipart_encode(data, files, boundary)
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        content_length = str(len(self.body))
+        content_type = self.content_type
+        return {"Content-Length": content_length, "Content-Type": content_type}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield self.body
+
+
+def encode(
+    data: RequestData = None,
+    files: RequestFiles = None,
+    json: typing.Any = None,
+    boundary: bytes = None,
+) -> RequestContent:
+    """
+    Handles encoding the given `data`, `files`, and `json`, returning
+    a `RequestContent` implementation which provides a byte iterator onto
+    the content, as well as `.is_rewindable()` and `.get_headers()` interfaces.
+
+    The `boundary` argument is also included for reproducible test cases
+    when working with multipart data.
+    """
+    if data is None:
+        if json is not None:
+            return JSONRequestContent(json)
+        elif files:
+            return MultipartRequestContent({}, files, boundary=boundary)
+        else:
+            return RequestContent()
+    elif isinstance(data, dict):
+        if files:
```

Comment:
For consistency/explicitness:

```suggestion
        if files is not None:
```

Reply:
I've commited the suggestion, yup. It does *somewhat* depend on exactly what behavior we want to expect here... ie should `files={}` result in a multipart encoded request, or a urlencoded request?

Reply:
Well, I'd say if `files` is given then as a user I'd expect HTTPX to use multipart. The previous behavior (which we still have in the `data is None` branch?) would use urlencoded for `data=..., files={}`, which I'd find surprising.

(But I'm also wondering whether using `files={}` has any practical sense in the first place?)